db/src/tx_checking.rs

```diff
@@ -44,3 +52,62 @@ where I: IntoIterator<Item=&'a AttributedTerm<'a>> {
 
     errors
 }
+
+/// Ensure that the given terms obey the cardinality restrictions of the given schema.
+///
+/// That is, ensure that any cardinality one attribute is added with at most one distinct value for
+/// any specific entity (although that one value may be repeated for the given entity).
+/// It is an error to:
+///
+/// - add two distinct values for the same cardinality one attribute and entity in a single transaction
+/// - add and remove the same values for the same attribute and entity in a single transaction
+/// - add any attribute to an entity that is also the subject of a `:db/retractEntity` term
+///
+/// We try to be maximally helpful by yielding every malformed set of datoms, rather than just the
+/// first set, or even the first conflict.  In the future, we might change this choice, or allow the
+/// consumer to specify the robustness of the cardinality checking desired.
+pub(crate) fn cardinality_conflicts<'a, I>(terms: I) -> Vec<CardinalityConflict>
+where I: IntoIterator<Item=&'a AttributedTerm<'a>> {
+    let mut errors = vec![];
+
+    let mut added: BTreeMap<Entid, BTreeMap<(Entid, bool), BTreeSet<&TypedValue>>> = BTreeMap::default();
```

Comment:
You're basically building a trie here. If you slightly tweak your approach, at this point you can:

1. Look up attributes, find the value type once, and make sure each incoming value is of the right type.
2. Implement the one/many distinction in two tries, using the presence of an `Entry` as a cardinality failure.
3. Drop `Term`s at this point and use the tries instead.

It seems wasteful to do all of this work and then not use the output for other things!

Also: it would be worth considering whether a `BTreeMap` and `BTreeSet` are the right data structures to use. Perhaps switch between `Hash*` and `BTree*` and use `cargo cli` to `.i fixtures/cities.schema`, `.i fixtures/all_seattle.edn` and see which is faster? I suspect the hash versions will be better.